### PR TITLE
Enable passing --xcoms to tasks test CLI command

### LINKED ISSUE #24232: Airflow tasks test is broken for xcom_pull based tasks
### Apache Airflow version

2.3.2 (latest released)

### What happened

Testing example_xcom_args via CLI throws following exception.
````
airflow.exceptions.AirflowException: XComArg result from generate_value at example_xcom_args with key="return_value" is not found!

full exception stack:
[2022-06-05 08:46:48,081] {taskinstance.py:1889} ERROR - Task failed with exception
Traceback (most recent call last):
  File "/venv-af/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1451, in _run_raw_task
    self._execute_task_with_callbacks(context, test_mode)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 1555, in _execute_task_with_callbacks
    task_orig = self.render_templates(context=context)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/taskinstance.py", line 2212, in render_templates
    rendered_task = self.task.render_template_fields(context)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/baseoperator.py", line 1185, in render_template_fields
    self._do_render_template_fields(self, self.template_fields, context, jinja_env, set())
  File "/venv-af/lib/python3.9/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/abstractoperator.py", line 344, in _do_render_template_fields
    rendered_content = self.render_template(
  File "/venv-af/lib/python3.9/site-packages/airflow/models/abstractoperator.py", line 398, in render_template
    return tuple(self.render_template(element, context, jinja_env) for element in value)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/abstractoperator.py", line 398, in <genexpr>
    return tuple(self.render_template(element, context, jinja_env) for element in value)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/abstractoperator.py", line 394, in render_template
    return value.resolve(context)
  File "/venv-af/lib/python3.9/site-packages/airflow/utils/session.py", line 71, in wrapper
    return func(*args, session=session, **kwargs)
  File "/venv-af/lib/python3.9/site-packages/airflow/models/xcom_arg.py", line 152, in resolve
    raise AirflowException(
airflow.exceptions.AirflowException: XComArg result from generate_value at example_xcom_args with key="return_value" is not found!
````

### What you think should happen instead

It should have return the results without any exception. 

However, test dags works without exceptions. 
````
airflow dags test example_xcom_args 2022-06-05
````

It seems that can't test individual tasks via CLI if the task has xcom.  I think, since airflow 2.3.x, xcom stores the task results in-memory instead of DB, it couldn't retrieve previous task result from xcom.   

### How to reproduce

Run tasks from airflow example_dags/example_xcomargs.py or example_xcom.py

````
airflow tasks test example_xcom_args generate_value 2022-06-05
airflow tasks test example_xcom_args print_value 2022-06-05
````

### Operating System

Mac OS and Python 3.9.12 

### Versions of Apache Airflow Providers

_No response_

### Deployment

Other

### Deployment details

_No response_

### Anything else

I believe it is a regression since airflow 2.3.x release xcom changes.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #28287: Testing tasks from CLI when they need dependencies from other tasks
UPDATE: The title of the issue was updated (and type was changed to a feature) to reflect the discussion that we need a feature where we can supply dependencies from other tasks to those run via `task test` CLI command.

### Apache Airflow version

2.5.0

### What happened

Let's have following DAG with task `echo`:

```python
@dag(start_date=pendulum.now())
def playground():

    @task
    def echo(message: str):
        logging.info(f'Received message: "{message}"')
        return message

    echo("hello world")

playground()
```

I would like to test the task from CLI. So according to the help and documentation I've created JSON with required parameter for the function and then I ran the following command:

```bash
$ airflow tasks test playground echo -t '{"message": "this is airflow"}'
```
In the output I'll get the following lines:

```
[2022-12-10 21:15:19,874] {playground.py:10} INFO - Received message: "hello world"
[2022-12-10 21:15:19,874] {python.py:177} INFO - Done. Returned value was: hello world
```

In my opinion, this is wrong behavior. 

### What you think should happen instead

If I provide the content of `message` parameter for the task, it should be used. So instead of message `"hello world"` the message `"this is airflow"` should be displayed and returned. Also it looks like, the task was called with the parameter from inside of the DAG.

If I change the task this way:

```python
@task
def echo(message: str, *args, **kwargs):
   print(kwargs['params'])
   logging.info(f'Received message: "{message}"')
   return message
```

I can see in the output, that the parameter `message` from the test was really passed to the task with the proper value:

```
{'message': 'this is airflow'}
[2022-12-10 21:26:10,495] {playground.py:11} INFO - Received message: "hello world"
[2022-12-10 21:26:10,495] {python.py:177} INFO - Done. Returned value was: hello world
```

### How to reproduce

1. Write simple task with TaskFlow API, such this:

   ```python
   import logging
   from airflow.decorators import task, dag
   import pendulum
  
  
   @dag(start_date=pendulum.now())
   def playground():
       @task
       def echo(message: str, *args, **kwargs):
           print(kwargs['params'])
           logging.info(f'Received message: "{message}"')
           return message
      
       echo("hello world")
  
  
   playground()
   ```

2. Test the task from CLI:

   ```bash
   $ airflow tasks test playground echo -t '{"message": "this is airflow"}'
   ```

### Operating System

Fedora Linux 37

### Versions of Apache Airflow Providers

_No response_

### Deployment

Virtualenv installation

### Deployment details

_No response_

### Anything else

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by vemikhaylov**
### Context

During discussion under #28287 it was decided that it may've been useful to be able to pass XCom Args to a task, when it's being tested with the `tasks test` CLI command and notify a user if a task is dependent on some XCom Arg, but it hasn't been passed.

### Solution

Introducing the `--xcoms` argument for the `tasks test` command, which a JSON list with XCom objects can be passed to. The JSON has the following structure: `[{"task_id": "my_task", "value": "foo"}, {"task_id": "another_task", "key": "custom_key", "value": 42}]`, which allows user to work both with the cases of the implicit XCom used by the Task Flow API (i.e., `xcom_key == "return_value"`) and of the XCom args with the custom keys (e.g., manually pushed and / or used in the templated kwargs). If `"key"` is omitted for any of the XCom object, `"return_value"` is used for it by default. The format of the passed argument is validated with the corresponding JSON schema.

Hence, a full CLI call may look like: 

`airflow tasks test example_passing_xcoms_via_test_command python_echo --xcoms [{"task_id": "get_python_echo_message", "value": "test xcom arg"}]`

or  

`airflow tasks test example_passing_xcoms_via_test_command run_bash --xcoms [{"task_id": "get_bash_command", "key": "return_value", "value": "echo \'test bash\'"}]`

If a user misses any XCom args in their call, they'll be able to see a message like:

```
airflow.exceptions.AirflowException: The task 'python_echo' is dependent on XComs passed from the upstream tasks. The following XComs are currently missed: [{"task_id": "get_python_echo_message", "key": "return_value"}]. Please, pass them via the --xcoms argument (see --help for more details).
```

closes: #28287
closes: #24232

### CODE REVIEW

**Comment by uranusjr** [comment]
Using JSON for this feels cumbersome to me. I don’t have a much better idea though. Also we need to find a way to support custom backends that may not store values as JSON.

**Comment by vemikhaylov** [comment]
@uranusjr thank you for your feedback!

> Using JSON for this feels cumbersome to me.

I think it does for me as well.. At the same time, it's probably worth mentioning that `--env-vars` and `--task-params` also already use JSON for the command, so it might be not a bad idea to have `--xcom-args` consistent with them.

Thinking out loud which alternative options of passing things through we may also have.

**1. Simplified format in CLI**

(task_id, xcom_key, value) triplets joined with a delimiter (a comma below).

```
--xcom-args get_python_echo_message,return_value,"test xcom arg" manually_push_bash_command,command,"echo \'test bash\'"
```

Probably all the escaping and ad-hoc handling will make things even worse.

**2. Values passed through a file**

```
--xcom-args task_args.json
```

Sorry for ".json" - Yeah, like again, the contents of the file should have some structure. We may go with yaml :)

What generally doesn't look convenient here with whatever structure of the file - requiring to create a file even for simple scenarios (if passing through the file is the only option). 

--
> Also we need to find a way to support custom backends that may not store values as JSON.

Let me clarify to make sure that I understand it correctly. You mean that generally the `BaseXCom` model allow us to store any binary (`LargeBinary`) `value`'s and not all of them may be passed as JSON and through CLI itself? Probably for that the approach #2 above can be used like: 

```
--xcom-args task_args.pickle
```

Though from the usability perspective I wouldn't stick to this option as the only available one. What do you think?

**Comment by dimberman** [comment]
Hi @vemikhaylov thank you for setting up this PR. This is a pretty important missing piece of the local dev story.

It would make sense also to offer the option to pass in a json/yaml file path. For a more complex item I imagine this would make more sense then needing to manually add the json object in the CLI arguments.

Another question is whether we should allow people to point at pickle files for pickle objects. I know we generally discourage people from turning on pickle, but it is a thing people do so we might want to allow those folks to locally test.

cc: @uranusjr WDYT?

**Comment by uranusjr** [comment]
Allowing a path makes sense to me. Since a JSON string has a pretty distinctive shape (starts and ends with either brackets or braces depending on which format we decide to use), I think we can allow both with one option.

While path to a pickle object also makes sense in theory, generating that pickle file would be pretty cumbersome and I’m not sure people would want to use that. I would probably leave that out for now until someone asks for it.

**Comment by uranusjr** [comment]
By the way, the argument name can probably be shortened to just `--xcom` or `--xcoms`?

**Comment by vemikhaylov** [comment]
> `--xcoms`

I love it! Though technically the "task" declaration operates with the `XComArg` layer, inside the task execution it may call `xcom_pull` and `xcom_push` directly, which is about pure `XCom`, so it may be fine and even more precise to go laconic.  

> ```
> airflow task run ... \
>     --xcom task1 key value \
>     --xcom task2 k1 v1 \
>     --xcom task2 k2 v2
> ```
> 
> I think this is possible by setting `nargs=3`.

It looks really great! Though there are the following issues, which I see with the format:

1. It's not rather extensible. Imagine we want to pass one more argument (e.g., `map_index` - I'm not sure if it's applicable, but let's imagine it's). Then it becomes `--xcom task1 key value map_index` - may be becomes less intuitive interface, where you should remember the order of the arguments.

2. It's possible to introduce the default key (i.e., `return_value`) by interpreting 2 args as `--xcom task1 value`, but it makes  even harder to extend it to more args in the future if needed.

3. It's not consistent with `--env-vars` and `--task-params`, which are JSON already.

Please let me know if those are valid concerns. I personally prefer more explicit interfaces and sometimes lean towards overly explicit ones, so it all may be rather subjective, feel free to push back :)

But so far I would probably stick to:

```
--xcoms [{"task_id": "get_python_echo_message", "value": "test xcom arg"}, ...]
```
and
```
--xcoms my_xcoms.json
```

In terms of execution, I wanted to ask if it's fine if I move iteratively here and break it down into two PRs: 1) CLI value passing; 2) File value passing? Among other things, it should facilitate the review process a bit. The first part should be self-contained as well, so probably a safe change to be committed separately?

**Comment by uranusjr** [comment]
I think your plan makes sense. So we’ll use this PR to implement value-passing, and add path support in a later PR.

File: airflow/cli/cli_parser.py [hunk 1]
```diff
@@ -36,13 +38,24 @@
 from airflow.exceptions import AirflowException
 from airflow.executors.executor_constants import CELERY_EXECUTOR, CELERY_KUBERNETES_EXECUTOR
 from airflow.executors.executor_loader import ExecutorLoader
+from airflow.models import XCOM_RETURN_KEY
```
Comment (vemikhaylov) [inline airflow/cli/cli_parser.py]: Are we fine with this dependency on the model layer here or it's better to refactor the constant out?

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -2741,6 +2749,14 @@ def tg2(inp):
         map_index_start = ancestor_map_index * further_count
         return range(map_index_start, map_index_start + further_count)
 
+    def inject_xcoms(self, xcoms: list[dict[str, Any]]):
+        """
+        To inject upstream dependencies' output instead of trying
+        to read from DB when testing a task individually.
+        """
```
Comment (vemikhaylov) [inline airflow/models/taskinstance.py:2774]: To be honest, I'm not quite sure about the approach, so would ask for feedback. A potential risk and questions around the current one:

1. The users of the framework may start to use it as a public API, but we may want to change how it works at some point in the future. Is it a valid concern or not?

2. Can the users use `XCom` directly without the API's exposed by `TaskInstance` or the models aren't available in the DAGs, so we cover the scenarios of the interactions with `XCom` on the `TaskInstance` layer? 

--

Alternative options, which I thought of:

1. Pass `xcoms` down the call stack, like to `ti.run(injected_xcoms=xcoms)`. Extending the `ti.run` signature didn't  look quite attractive to me, though may be more clear and explicit in comparison to preliminary "injection". But injection is similar to what we do with `params`: https://github.com/apache/airflow/blob/4e3b5ae7248c2327864f64b25dc7a5bd7705430c/airflow/cli/commands/task_command.py#L590

2. Have a module with a global variable, which we could initialise in `task_command` and then read from either in `TaskInstance.xcom_pull` or `BaseXCom.get_one` / `.get_many`. Global state..

3. Monkeypatch `ti.xcom_pull` (dirty?)

4. Implement a custom `XCom` backend to be used in `xcom.py`, which would read the injected data from memory. 

5. If we are able to run everything inside a common transaction, which is rolled-back at the end of the command execution, then we can just "put" the corresponding XComs in "DB" before running the task?

--

Since I wasn't sure about feasibility and validity of some of the approaches above (especially 4 and 5), I implemented the most straightforward way, similar to what we already have (the `params` reference above), seeking for feedback.

File: airflow/cli/commands/task_command.py [hunk 3]
```diff
@@ -596,6 +597,10 @@ def task_test(args, dag=None):
         task, args.map_index, exec_date_or_run_id=args.execution_date_or_run_id, create_if_necessary="db"
     )
 
+    xcoms = args.xcoms or []
+    _validate_injected_xcoms(xcoms, task)
```
Comment (vemikhaylov) [inline airflow/cli/commands/task_command.py:602]: If we stick to the `ti.inject_xcoms` approach, then it may be considered moving the validation logic to `ti.inject_xcoms` itself. It'll imply that if `inject_xcoms` is called, it should inject all the required XComs at once, but probably that's fine.

Also the exception raised from there will lack the `"Please, pass them via the --xcoms argument (see --help for more details)."` part (in the context of the method we won't know where it's called from). It will need either to be caught and added in `task_command` or omitted.

File: airflow/example_dags/example_passing_xcoms_via_test_command.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    bash_push = BashOperator(
+        task_id="manually_push_bash_command",
+        bash_command='{{ ti.xcom_push(key="command", value="echo \'default\'") }}',
+    )
+    BashOperator(
+        task_id="run_bash_with_manually_pushed_command", bash_command=XComArg(bash_push, key="command")
+    )
+    BashOperator(
+        task_id="run_bash_with_manually_pulled_command",
+        bash_command='{{ task_instance.xcom_pull(task_ids="manually_push_bash_command", key="command") }}',
```
Comment (vemikhaylov) [inline airflow/example_dags/example_passing_xcoms_via_test_command.py:53]: We don't currently cover this case with the beforehand validation of the passed XComs since `XComArg.iter_xcom_references(task)` won't include it, it's a lower level. We may try to identify it real-time, when it's being queried to return a meaningful error.

But generally DAG doesn't identify it as a dependency either, because it seems to be hard to do, so probably it's relatively acceptable and this use case is mostly on the user?

File: airflow/cli/cli_parser.py (hunk 1 above)
Comment (uranusjr) [reply airflow/cli/cli_parser.py]: Better to refactor the constant out.

File: airflow/cli/commands/task_command.py (hunk 3 above)
Comment (uranusjr) [reply airflow/cli/commands/task_command.py:602]: > Also the exception raised from there will lack the [context]

You can raise something other than AirflowException (I don’t see why you need to raise that exception), and catch it here to provide the additional context.

File: airflow/models/baseoperator.py [hunk 5]
```diff
@@ -1119,6 +1121,12 @@ def set_xcomargs_dependencies(self) -> None:
                 arg = getattr(self, field)
                 XComArg.apply_upstream_relationship(self, arg)
 
+    def iter_xcom_dependencies(self) -> Iterator[tuple[Operator, str]]:
+        """Upstream dependencies that provide XComs used by this operator."""
+        from airflow.models.xcom_arg import XComArg
+
+        yield from XComArg.iter_xcom_references(self)
```
Comment (uranusjr) [inline airflow/models/baseoperator.py:1141]: Why is this needed?

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (uranusjr) [reply airflow/models/taskinstance.py:2774]: > Can the users use XCom directly without the API

They can, via `XCom.get_one` or `get_many`. So I feel the only fully reliable way is to actually write the data to the database so they are available directly from the XCom interface.

We already do something similar for DagRun (if you specify a DagRun that does not actually exist, we would create a temporary one in the database for the `test` run and delete it afterwards), so I can think of a couple of approaches:

1. Write the custom XCom data to a separate table, and read them on `XCom`. Delete them after the test run.
2. Write the custom data to the real XCom table, and delete them afterwards. Emit an error (and refuse to execute) if there is already existing data matching the custom input in the table. This feels reasonable since it would only be possible if you test a task _in an existing DAG run but don’t want to use the actual upstream XComs_, which feels very niche and not a necessary feature to me.

File: airflow/cli/cli_parser.py (hunk 1 above)
Comment (vemikhaylov) [reply airflow/cli/cli_parser.py]: Created it as a separate PR to keep this one more focused (#30180).

File: airflow/cli/commands/task_command.py (hunk 3 above)
Comment (vemikhaylov) [reply airflow/cli/commands/task_command.py:602]: > I don’t see why you need to raise that exception

Let me clarify if you are questioning why we should raise an exception at all (like not go with returning a bool flag or something) or you're not sure if we need the `AirflowException` class here? If it's the latter, what is the intended scenario for that? I just noticed that it had been used for the args validation in the module.

But generally absolutely aligned! Actually I also mentioned that it may've been intercepted and enriched, definitely not a blocker, just a consideration.

File: airflow/models/baseoperator.py (hunk 5 above)
Comment (vemikhaylov) [reply airflow/models/baseoperator.py:1141]: Is it a question why we need to define a method for that instead of calling `XComArg.iter_xcom_references` directly? As I see that, when we look at a `task` and want to understand, which XComs it depends on, `XComArg.iter_xcom_references(task)` seems to me like a lower level detail of implementation. Something similar happens in the mapped operator:

https://github.com/apache/airflow/blob/3f6b5574c61ef9765d077bdd08ccdaba14013e4a/airflow/models/mappedoperator.py#L638-L643

File: airflow/models/taskinstance.py (hunk 2 above)
Comment (vemikhaylov) [reply airflow/models/taskinstance.py:2774]: > Write the custom data to the real XCom table, and delete them afterwards. Emit an error (and refuse to execute) if there is already existing data matching the custom input in the table. This feels reasonable since it would only be possible if you test a task in an existing DAG run but don’t want to use the actual upstream XComs, which feels very niche and not a necessary feature to me.

For this one I have a concern that in general script may exit in a tricky way, even skipping `finally` (https://stackoverflow.com/questions/49262379/does-finally-always-execute-in-python). When we're dealing with a temporary DAG run, then that's relatively fine and is visible from the run id:

https://github.com/apache/airflow/blob/main/airflow/cli/commands/task_command.py#L74-L80

But for a normal DAG run it may leave the data in a strange state. Also probably if this is an ongoing DAG run, then, if I'm not mistaken, there might be race conditions like "write XCom in test command" -> "write the same XCom in a real task executed in the DAG run" -> "delete XCom in test command (clean up)".

What do you think?

File: airflow/models/baseoperator.py (hunk 5 above)
Comment (uranusjr) [reply airflow/models/baseoperator.py:1141]: `iter_xcom_references` is an internal interface that Airflow internals are free to use. MappedOperator has a wrapper for it because the wrapper also references a private member; this is not needed for BaseOperator.

**Comment by potiuk** [comment]
FYI: Fix to failing test in https://github.com/apache/airflow/pull/29608

**Comment by potiuk** [comment]
You should rebase now.

File: airflow/models/baseoperator.py (hunk 5 above)
Comment (vemikhaylov) [reply airflow/models/baseoperator.py:1141]: Is the recommendation to call `XComArg.iter_xcom_references` directly then, right?
